### ISSUE TITLE
[Clang] Dynamic builtin_shufflevector crahes with vector of bool as indices

### ISSUE THREAD

**Issue by AmrDeveloper** [llvm/llvm-project#218132]
In the builtin_shufflevector dynamic shuffling mode, it accepts two vectors and uses the second vector as indices.

We mask the indices using `AND splat (llvm::NextPowerOf2(LHSElts - 1) - 1)` on each vector element

https://github.com/llvm/llvm-project/blob/8f5ef7362dc9dbb19454e458f7ca4fb99fab5d52/clang/lib/CodeGen/CGExprScalar.cpp#L2025-L2027

The assert was triggered because the masked value can't be fit in the i1 type

CompilerExplorer: https://godbolt.org/z/zEz8ors95

```
clang++: /root/llvm-project/llvm/include/llvm/ADT/APInt.h:127: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isUIntN(BitWidth, val) && "Value is not an N-bit unsigned value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and dumped files.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/clang++ -g -o /app/output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-snapshot -fcolor-diagnostics -fno-crash-diagnostics -emit-llvm <source>
1.	<eof> parser at end of file
2.	<source>:3:6: LLVM IR generation of declaration 'vec_bool_logical_not'
3.	<source>:3:6: Generating code for declaration 'vec_bool_logical_not'
 #0 0x00000000044143e8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x44143e8)
 #1 0x0000000004411294 llvm::sys::RunSignalHandlers() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4411294)
 #2 0x00000000044118cc llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x44118cc)
 #3 0x000000000434ff58 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #4 0x00007013f0e45330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #5 0x00007013f0e9ec0c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x9ec0c)
 #6 0x00007013f0e4527e raise (/lib/x86_64-linux-gnu/libc.so.6+0x4527e)
 #7 0x00007013f0e288ff abort (/lib/x86_64-linux-gnu/libc.so.6+0x288ff)
 #8 0x00007013f0e2881b (/lib/x86_64-linux-gnu/libc.so.6+0x2881b)
 #9 0x00007013f0e3b517 (/lib/x86_64-linux-gnu/libc.so.6+0x3b517)
#10 0x0000000000e1c579 llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xe1c579)
#11 0x0000000003be3be5 llvm::ConstantInt::get(llvm::Type*, unsigned long, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3be3be5)
#12 0x000000000497cdf2 (anonymous namespace)::ScalarExprEmitter::VisitShuffleVectorExpr(clang::ShuffleVectorExpr*) (.isra.0) CGExprScalar.cpp:0:0
#13 0x000000000497008e clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) CGExprScalar.cpp:0:0
#14 0x000000000497586c clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x497586c)
```

- We accept a vector of bools as indices because it has an int representation, but do we actually want that?

https://github.com/llvm/llvm-project/blob/0d18bb67d106cdf1e2c9e114ac1b67167a233b6c/clang/lib/Sema/SemaChecking.cpp#L6588-L6590

- If we accept a vector of bools as indices, can we eliminate the mask if the first vector's size is > 1, because all indices are guaranteed to be 0 or 1 🤔

**Comment by AmrDeveloper** [comment]
cc @AaronBallman @jhuber6

**Comment by jhuber6** [comment]
What is the expected behavior here? Bools have an integral relationship but it doesn't make much sense to use them as an index. You could easily make this work by extending it to a standard integer type, but then what? You'd only ever be able to index two values. I'd say this should probably just be a compiler error instead of an assertion.

**Comment by AmrDeveloper** [comment]
> Bools have an integral relationship but it doesn't make much sense to use them as an index. You could easily make this work by extending it to a standard integer type, but then what? You'd only ever be able to index two values

Yes, creating a workaround is easy, but it's about whether it's useful and makes sense, and also about what we expect from this built-in.

I think it's a good idea to create a compiler error when indices don't make sense

**Comment by AaronBallman** [comment]
> I'd say this should probably just be a compiler error instead of an assertion.

I think that makes sense as well. We can relax the compile error if someone comes forward with a sufficiently compelling use case.

**Comment by AmrDeveloper** [comment]
Great, I will implement it soon

**Comment by AmrDeveloper** [comment]
Another improvement from the same diff is this case

```c
typedef float v8f __attribute__((ext_vector_type(8)));

void vector_of_float_mask() {
  v8f a;
  v8f b;
  auto r = __builtin_shufflevector(a, b);
}
```

Previously it reports

```
error: first two arguments to '__builtin_shufflevector' must have the same type
```

They already have the same type :D.

https://godbolt.org/z/cEvPEdT33
